loop customization docs (#9557)
PyTorchLightning/pytorch-lightning: docs/source/advanced/loops.rst

By advanced users we mean users that are familiar with the major components under the Trainer and how they interact with the LightningModule.

In this advanced user guide we will learn about

- how the Trainer runs a loop,
- the Loop base class,
- the default loop implementations and subloops Lightning has,
- how Lightning defines a tree structure of loops and subloops,
- how you can create a custom loop for a new training step flavor,
- and how you can connect the custom loop and run it.

Most importantly, we will also provide guidelines when to use loop customization and when NOT to use it.

Trainer entry points for loops
------------------------------

The Trainer has four entry points for training, testing and inference, and each method corresponds to a main loop:

+---------------------------------------------------------------+-----------------------------------------------------------------------+-------------------------------------------------------------------------------+

> | Entry point                                                   | Trainer attribute                                                     | Loop class                                                                    |

Author: this is supposed to be a table, sorry this may look messed up in your browser when viewing the diff

Reviewer: > this may look messed up in your browser

For better text readability, we could use `.. list-table::` as we do in Bolts maybe?

source: https://github.com/PyTorchLightning/lightning-bolts/blame/master/docs/source/deprecated/models/self_supervised.rst#L148
built page: https://lightning-bolts.readthedocs.io/en/latest/deprecated/models/self_supervised.html#cifar-10-and-stl-10-baselines

Author: looks better on the source code. let's do it

.. _loop_customization:

Loop Customization
==================

> Loop customization is an experimental feature introduced in Lightning 1.5 that enables advanced users to write new training logic or modify existing behavior in Lightning's training-, evaluation-, or prediction loops.

Reviewer 2: nit: the `-` notation is not very commonly known, I wouldn't put it in docs

Author: ok will remove, I'm not sure if it's correct without them.

> By advanced users we mean users that are familiar with the major components under the Trainer and how they interact with the LightningModule.

Reviewer 2: ```suggestion
By advanced users, we mean users that are familiar with the major components under the ``Trainer`` and how they interact with the ``LightningModule``.
```

> In this advanced user guide we will learn about

Reviewer 2: ```suggestion
In this advanced user guide, we will learn about:
```

> - how the Trainer runs a loop,

Reviewer 2: ```suggestion
- how the ``Trainer`` runs a loop
```

> - the Loop base class,

Reviewer 2: ```suggestion
- the ``Loop`` base class design,
```

> - the default loop implementations and subloops Lightning has,

Reviewer 2: shouldnt each bullet point start with a capital and end with a period?

Author: it's part of one large sentence. 
individually, these wouldn't be sentences

Reviewer 3: Agree with @carmocca this doesn't work well imo. Could we make these full sentences and bullet point them?

- how Lightning defines a tree structure of loops and subloops,

> - how you can create a custom loop for a new training step flavor,

Reviewer 2: ```suggestion
- how you can create a custom loop for a new `training_step` flavor,
```

- and how you can connect the custom loop and run it.

> Most importantly, we will also provide guidelines when to use loop customization and when NOT to use it.

Reviewer 2: ```suggestion
Most importantly, we will also provide guidelines on when to use loop customization and when NOT to use it.
```

+---------------------------------------------------------------+-----------------------------------------------------------------------+-------------------------------------------------------------------------------+
| :meth:`~pytorch_lightning.trainer.trainer.Trainer.test`       | :attr:`~pytorch_lightning.trainer.trainer.Trainer.test_loop`          | :class:`~pytorch_lightning.loops.dataloader.evaluation_loop.EvaluationLoop`   |
+---------------------------------------------------------------+-----------------------------------------------------------------------+-------------------------------------------------------------------------------+
| :meth:`~pytorch_lightning.trainer.trainer.Trainer.predict`    | :attr:`~pytorch_lightning.trainer.trainer.Trainer.predict_loop`       | :class:`~pytorch_lightning.loops.dataloader.prediction_loop.PredictionLoop`   |
+---------------------------------------------------------------+-----------------------------------------------------------------------+-------------------------------------------------------------------------------+

> When the user calls :code:`Trainer.method`, it redirects to the corresponding :code:`Trainer.loop.run()` which implements the main logic of that particular Lightning loop.

Reviewer 2: The table uses entry point
```suggestion
When the user calls :code:`Trainer.<entry-point>`, it redirects to the corresponding :code:`Trainer.loop.run()` which implements the main logic of that particular Lightning loop.
```

Think of it as the start of a Python :code:`while` loop.
The :meth:`~pytorch_lightning.loops.base.Loop.run` method is part of the base :class:`~pytorch_lightning.loops.base.Loop` class that every loop inherits from (like every model inherits from LightningModule).

The Loop base class
-------------------

The :class:`~pytorch_lightning.loops.base.Loop` class is the base for all loops in Lighting just like the LightningModule is the base for all models.
It defines a public interface that each loop implementation must follow, the key ones are:

- :meth:`~pytorch_lightning.loops.base.Loop.advance`: implements the logic of a single iteration in the loop
- :meth:`~pytorch_lightning.loops.base.Loop.done`: a boolean stopping criteria
- :meth:`~pytorch_lightning.loops.base.Loop.reset`: implements a mechanism to reset the loop so it can be restarted

> These methods are called by the default implementation of the :meth:`~pytorch_lightning.loops.base.Loop.run` entry point as shown in the code excerpt below.

Reviewer 2: ```suggestion
These methods are called by the default implementation of the :meth:`~pytorch_lightning.loops.base.Loop.run` entry point as shown in the (reduced) code excerpt below.
```

.. code-block:: python

    def run(self, *args, **kwargs):

        self.reset()
        self.on_run_start(*args, **kwargs)

        while not self.done:
            try:
                self.advance(*args, **kwargs)
            except StopIteration:
                break

        output = self.on_run_end()
        return output

> Some important observations here: One, the `run()` method can define input arguments that get forwarded to some of the other methods that get invoked as part of `run()`.

Reviewer 2: ```suggestion
Some important observations here: One, the ``run()`` method can define input arguments that get forwarded to some of the other methods that get invoked as part of ``run()``.
```

> Such input arguments typically comprise of one or several iterables over which the loop is suppose to iterate, for example, an iterator over a :class:`~torch.utils.data.DataLoader`.

Reviewer 2: ```suggestion
Such input arguments typically comprise of one or several iterables over which the loop is supposed to iterate, for example, an iterator over a :class:`~torch.utils.data.DataLoader`.
```

The reason why the inputs get forwarded is mainly for convenience but implementations are free to change this.

> Secondly, `advance()` can raise a :class:`StopIteration` to exit the loop early.

Reviewer 2: ```suggestion
Secondly, ``advance()`` can raise a :class:`StopIteration` to exit the loop early.
```

> This is analogeous to a :code:`break` statement in a raw Python `while`-loop for example.

Reviewer 2: ```suggestion
This is analogous to a :code:`break` statement in a raw Python ``while`` for example.
```

> Finally, a loop may return an output as part of `run()`.

Reviewer 2: ```suggestion
Finally, a loop may return an output as part of ``run()``.
```

This output could for example be a list containing all results produced in each iteration (advance) of the loop.

> Loops can also be nested! That is, a loop may call another one inside of its `advance()`.

Reviewer 2: ```suggestion
Loops can also be nested! That is, a loop may call another one inside of its ``advance()``.
```

Default loop implementations
----------------------------

The training loop in Lightning is called *fit loop* and is actually a combination of several loops.
Here is what the structure would look like in plain Python:

.. code-block:: python

>     # FitLoop

Author: here it won't work because it's under a code-block

>     # FitLoop

Reviewer 4: Yes, but at would make search simpler within the codebase from this abstraction. Up to you :)

.. code-block:: python

    def training_step(self, batch, batch_idx):
        # do something with optimizer 0
        loss0 = self.loss(self(batch))
        yield loss0

        # do something with optimizer 1 that requires loss0
        loss1 = self.foo(loss0)
        yield loss1

Normally, we would implement a training step with a signature :code:`training_step(self, batch, batch_idx, optimizer_idx)` and then conditionally compute either :code:`loss0` or :code:`loss1` depending on the current optimizer index and return that loss at the end.
But if the computation of say :code:`loss1` depends on :code:`loss0` or another quantity computed for the first optimizer, we would have to recompute the value for :code:`optimizer_idx = 1` and that is wasteful.

With the training step as a generator as shown above however, we are able to retain the local variables across training_step boundaries when we switch from one optimizer to the next.
The alternative to this would be *manual optimization* where the same can be achieved, but with the generator loop we can still get all benefits of manual optimization without having to call backward or zero grad ourselves.

> In order to enable returning a generator from a training step, we need a custom loop!

Reviewer 4: ```suggestion
In order to enable returning a generator from a training step, we need to implement the custom yield loop!
```

> This will be a subclass of the existing :class:`~pytorch_lightning.loops.optimization.optimizer_loop.OptimizerLoop` and then be attached to the :class:`~pytorch_lightning.loops.batch.training_batch_loop.TrainingBatchLoop`.

Reviewer 4: ```suggestion
This will be a subclass of the existing :class:`~pytorch_lightning.loops.optimization.optimizer_loop.OptimizerLoop`. Once implemented, we would have to ``connect`` it to the :class:`~pytorch_lightning.loops.batch.training_batch_loop.TrainingBatchLoop`.
```

.. code-block:: python

    from functools import partial
    from pytorch_lightning.loops import Loop, OptimizerLoop
    from pytorch_lightning.loops.optimization.optimizer_loop import ClosureResult
    from pytorch_lightning.loops.utilities import _build_training_step_kwargs

>     class YieldLoop(OptimizerLoop):

Reviewer 4: Beautiful ! Lot of work to get there :)

>     class YieldLoop(OptimizerLoop):

Reviewer 4: @awaelchli, would mind making sure the code isn't too long. Maybe 80 lines. It doesn't render very well in the doc.

<img width="806" alt="Screenshot 2021-09-18 at 12 18 17" src="https://user-images.githubusercontent.com/12861981/133886985-d6cefec4-31eb-479b-8826-c8dc06dafbe4.png">

        def __init__(self):
            super().__init__()
            self._training_step_generator = None

        def connect(self, **kwargs):

>             raise NotImplementedError(f"{self.__class__.__name__} does not connect any child loops.")

Reviewer 4: ```suggestion
            raise NotImplementedError(f"{self.__class__.__name__} is a leaf loop and shouldn't be connected to a child. ")
```

Reviewer 4: quick note: I wonder if we want to show this :)

Author: yeah, maybe not. we can leave it out for brevity.

        def on_run_start(self, batch, optimizers, batch_idx):
            super().on_run_start(batch, optimizers, batch_idx)
            assert self.trainer.lightning_module.automatic_optimization

            # We request the generator once and save it for later so we can call next() on it.
            self._training_step_generator = self._get_training_step_generator(batch, batch_idx, opt_idx=0)

>         def _make_step_fn(self, batch, batch_idx, opt_idx):

Reviewer 4: Let's move utils at the bottom.

            return partial(self._training_step, self._training_step_generator)

        def _get_training_step_generator(self, batch, batch_idx, opt_idx):
            step_kwargs = _build_training_step_kwargs(
                self.trainer.lightning_module,
                self.trainer.optimizers,
                batch,
                batch_idx,
                opt_idx,
                hiddens=None,
            )

>             # Here we are basically calling lightning_module.training_step()
>             # and this returns a generator!

Reviewer 4: ```suggestion
            # Here we are basically calling lightning_module.training_step()
            # and this returns a generator! 
            # The training_step is handled by the accelerator to enable distributed training.
```

            generator = self.trainer.accelerator.training_step(step_kwargs)
            return generator

        def _training_step(self, training_step_generator):
            lightning_module = self.trainer.lightning_module

            with self.trainer.profiler.profile("model_forward"):
                lightning_module._current_fx_name = "training_step"
                with self.trainer.profiler.profile("training_step"):

                    # Here, instead of calling lightning_module.training_step()
                    # we call next() on the generator!
                    training_step_output = next(training_step_generator)
                    self.trainer.accelerator.post_training_step()

>                 training_step_output = self.trainer.call_hook("training_step_end", training_step_output)

Reviewer 4: ```suggestion
               # As you can see, ``self.trainer.call_hook`` enables calling 
               # any hook with this exact function name and provided arguments.
               # You can easily introduce new hooks within your custom loop.
                training_step_output = self.trainer.call_hook("training_step_end", training_step_output)
```

Reviewer 4: Would be nice to have an example where the users can create new hooks.

                result = ClosureResult.from_training_step_output(training_step_output, self.trainer.accumulate_grad_batches)
            return result

As we can see, not much work needs to be done to enable our generator training step.
The new loop is called :code:`YieldLoop` and contains a reference to the generator returned by the :code:`training_step`.

> On every new run (over the optimizers) we call the :code:`training_step` method on the LightningModule which is supposed to return a generator because it contains :code:`yield` statements.

Reviewer 4: ```suggestion
On every new run (over the optimizers) we call the :code:`training_step` method on the LightningModule which is supposed to return a generator as it contains the :code:`yield` statements.
```

There must be as many :code:`yield` statements as there are optimizers.

> Given this new loop, here is how you connect it to the Trainer:

Reviewer 4: ```suggestion
With the new yield loop, we can connect it to the :class:`~pytorch_lightning.loops.batch.training_batch_loop.TrainingBatchLoop` aka ``batch_loop`` and replace the current :class:`~pytorch_lightning.loops.optimization.optimizer_loop.OptimizerLoop`:
```

Reviewer 4: Feel free to decide to keep this one or the one below.

.. code-block:: python

    model = LitModel()
    trainer = Trainer()

    yield_loop = YieldLoop()
    trainer.fit_loop.epoch_loop.batch_loop.connect(optimizer_loop=yield_loop)

    trainer.fit(model)  # runs the new loop!

Note that we need to connect it to the :class:`~pytorch_lightning.loops.batch.training_batch_loop.TrainingBatchLoop` and we are replacing the default optimizer loop that Lightning provides.

Persisting the state of loops
-----------------------------

.. note::
    This is an experimental feature and is not activated by default.

>     Set the environment variable `PL_FAULT_TOLERANT_TRAINING = 1` to enable saving the progress of loops.

Reviewer 4: ```suggestion
    Set the environment variable ``PL_FAULT_TOLERANT_TRAINING = 1`` to enable saving the progress of loops.
```

>     Read more about :doc:`fault-tolerant training training <../advanced/fault_tolerant_training>`.

Reviewer 4: ```suggestion
    Read more about :doc:`fault-tolerant tolerant training <../advanced/fault_tolerant_training>`.
```

> An interesting property of the abstract loop interface is that it can maintain a state.

Reviewer 4: ```suggestion
The PyTorch Lightning Loop state management design is highly inspired from the PyTorch nn.Module and relies on the same underlying logic. 
```

Author: actually all I want to express here is that the benefit of having loops as objects is that they can have state.

Reviewer 4: I know, but I built the state management exactly as pytorch ones. Advanced users are already familiar with it, and it means they will understand faster how it works. Just a personal suggestion.

> It can save its state to a checkpoint through corresponding hooks and if implemented accordingly, resume it's state of exectuion at the appropriate place.

Reviewer 4: ```suggestion
As a result, the loops can save their state and their children within the model checkpoint through corresponding hooks and if implemented accordingly, resume their state of execution at the appropriate place.
```

> This design is particularly interesting for fault-tolerant training which is an experimental feature released in Lightning v1.5.

Reviewer 4: ```suggestion
As a matter of facts, all the Lightning loops implements an advanced state management which is the underlying mechanism of a long waited feature: Fault Tolerant Training which is an experimental feature released in Lightning v1.5.
```

The two hooks :class:`~pytorch_lightning.loops.base.Loop.on_save_checkpoint` and :class:`~pytorch_lightning.loops.base.Loop.on_load_checkpoint` function very similarly to how LightningModules and Callbacks save and load state.

.. code-block:: python

>     def on_save_checkpoint(self):
>         state_dict["iteration"] = self.iteration
>         return state_dict
> 
> 
>     def on_load_checkpoint(self, state_dict):
>         self.iteration = state_dict["iteration"]

Reviewer 4: ```suggestion
    def on_save_checkpoint(self):
        state_dict["iteration"] = self.iteration
        return state_dict

    def on_load_checkpoint(self, state_dict):
        self.iteration = state_dict["iteration"]
```

When the Trainer is restaring from a checkpoint (e.g., through :code:`Trainer(resume_from_checkpoint=...)`), the loop exposes a boolean :attr:`~pytorch_lightning.loops.base.Loop.restarting`.
Based around the value of this variable, the user can write the loop in such a way that it can restart from an arbitrary point given the state loaded from the checkpoint.
For example, the implementation of the :meth:`~pytorch_lightning.loops.base.Loop.reset` method could look like this given our previous example:

.. code-block:: python

    def reset(self):
        if not self.restarting:
            self.iteration = 0

FAQ
---

**Why are the loops in Lightning classes and not just simply `for` or `while` loops?**

Several reasons:
1) a loop can be portable, i.e., can be shared and imported in new projects,

> 2) managing state and being able to resume a loop is more easily realized with on object-oriented design,

Reviewer 4: ```suggestion
2) managing state and being able to resume a loop is more easily realized with an object-oriented design,
```

3) complex interactions between loops can be more easily modelled and leverage design patterns from object oriented design.

> **How do I make sure a given LightningModule is compatible with my custom loop?**

Reviewer 4: ```suggestion
**How do I make sure that a given LightningModule is compatible with my custom loop?**
```

> To restrict the compatibility of a LightningModule to a particular loop type, we recommend to define a specific class mixin for this purpose.

Reviewer 4: Dope ! Really neat :)
